pkg/kubelet/util/pod_startup_latency_tracker.go

```diff
@@ -0,0 +1,178 @@
+	// log, if pod latency was already Observed
+	metricRecorded bool
+}
+
+func NewPodStartupLatencyTracker() *PodStartupLatencyTracker {
+	return &PodStartupLatencyTracker{
+		pods:  map[types.UID]*perPodState{},
+		clock: clock.RealClock{},
+	}
+}
+
+// ObservedPodOnWatch to be called from somewhere where we look for pods.
+func (p *PodStartupLatencyTracker) ObservedPodOnWatch(pod *v1.Pod, when time.Time) {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	// pod Terminated and would never start
+	if pod.Status.Phase == v1.PodFailed {
+		delete(p.pods, pod.UID)
+		return
+	}
+
+	state := p.pods[pod.UID]
+	if state == nil {
+		// create a new record for pod, only if it was not yet acknowledged by the Kubelet
+		// this is required, as we want to log metric only for those pods, that where scheduled
+		// after Kubelet started
+		if pod.Status.StartTime.IsZero() {
+			p.pods[pod.UID] = &perPodState{}
+		}
+
+		return
+	}
+
+	if state.observedRunningTime.IsZero() {
+		// skip, pod didn't start yet
+		return
+	}
+
+	if state.metricRecorded {
+		// skip, pod's latency already recorded
+		return
+	}
+
+	if hasPodStartedSLO(pod) {
+		podStartingDuration := when.Sub(pod.CreationTimestamp.Time)
+		imagePullingDuration := state.lastFinishedPulling.Sub(state.firstStartedPulling)
+		podStartSLOduration := (podStartingDuration - imagePullingDuration).Seconds()
+
+		klog.InfoS("Observed pod startup duration",
+			"pod", klog.KObj(pod),
+			"podStartSLOduration", podStartSLOduration,
+			"pod.CreationTimestamp", pod.CreationTimestamp.Time,
+			"firstStartedPulling", state.firstStartedPulling,
+			"lastFinishedPulling", state.lastFinishedPulling,
+			"observedRunningTime", state.observedRunningTime,
+			"watchObservedRunningTime", when)
+
+		metrics.PodStartSLIDuration.WithLabelValues().Observe(podStartSLOduration)
+		state.metricRecorded = true
+	}
+}
+
+func (p *PodStartupLatencyTracker) RecordImageStartedPulling(podUID types.UID) {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	state := p.pods[podUID]
+	if state == nil {
+		return
+	}
+
+	if state.firstStartedPulling.IsZero() {
+		state.firstStartedPulling = p.clock.Now()
+	}
+}
+
+func (p *PodStartupLatencyTracker) RecordImageFinishedPulling(podUID types.UID) {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	state := p.pods[podUID]
+	if state == nil {
+		return
+	}
+
+	state.lastFinishedPulling = p.clock.Now() // Now is always grater than values from the past.
+}
+
+func (p *PodStartupLatencyTracker) RecordStatusUpdated(pod *v1.Pod) {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	state := p.pods[pod.UID]
+	if state == nil {
+		return
+	}
+
+	if state.metricRecorded {
+		// skip, pod latency already recorded
+		return
+	}
+
+	if !state.observedRunningTime.IsZero() {
+		// skip, pod already started
+		return
+	}
+
+	if hasPodStartedSLO(pod) {
+		klog.V(2).InfoS("Mark when the pod was running for the first time", "pod", klog.KObj(pod), "rv", pod.ResourceVersion)
+		state.observedRunningTime = p.clock.Now()
```

Comment:
I thought you wanted to record the time when we observed it running, not when the kubelet's internal state was updated?  Now() at this point is when the kubelet updates its internal state IIRC.

Reply:
That is correct and the observed time (the time of the watch) is what we use to calculate the Latency SLI:
https://github.com/kubernetes/kubernetes/pull/111930/files/785715519f2dbf539db3c84db044ae296fbcce51#diff-9365914435bee2af0a1f714a83aeb4b8c5ac8b1a4cf8ebe6e4fa83a5039ea31dR90

The `state.observedRunningTime` is the time "when the kubelet updates its internal state IIRC", exactly as you said - and it's used primarily for the debug / log. It might be used more in the future, ref: https://github.com/kubernetes/kubernetes/pull/111930#issuecomment-1223748694

Reply:
Thanks